lib/AST/Identifier.cpp

```diff
@@ -79,9 +96,29 @@ int Identifier::compare(Identifier other) const {
   return str().compare(other.str());
 }
 
+DeclName DeclName::createSubscript(ASTContext &C,
+                                   ArrayRef<Identifier> argumentNames) {
+  return DeclName(C, Identifier(SubscriptIdentifierData), argumentNames);
+}
+
+DeclName DeclName::createSubscript() {
+  return Identifier(SubscriptIdentifierData);
+}
+
+DeclNameKind DeclName::getKind() const {
```

Comment:
I think the problem was that if you initialise `SubscriptIdentifierData ` as a private static member on `DeclName` in the header of `DeclName` it gets a different value for each library that import `Identifier.h`. This worked around the issue.